### PR TITLE
feat: add vmTrace support

### PR DESCRIPTION
**Pull request by banteg**
### What I did

add support for `vmTrace` option in parity-style traces. this allows to get very similar data to geth debug traces, but it works an order of magnitude faster with the rpc responses being three orders of magnitude smaller.

### How I did it

unlike geth, which returns the storage and memory at each frame, this format returns incremental changes of which part of memory was overwritten or which items were pushed onto the stack, which is enough for us to recompute and return an iterator with up-to-date values at each frame at >200k frames/second.

i also added a map for opcodes consuming items from stack, since this info is not a part of the incremental updates.

i use py-evm primitives for an evm-compliant stack and memory implementations.

i also added the current contract address to the context of trace frame.

im using [msgspec](https://jcristharif.com/msgspec/) library here instead of pydantic since it's an order of magnitude faster. i suggest we replace pydantic in this package later on since we mostly work with quite heavy responses here.

### How to verify it

i've assembled a demo which can do pairwise comparison with geth here https://gist.github.com/banteg/7e81dea4505ea07d34230001a6868ac4

`evm_trace.vmtrace.to_trace_frames` can be used as a **drop-in replacement for geth traces**, but it's 1-2 orders of magnitude faster.

### References
- https://docs.rs/web3/latest/web3/types/struct.VMTrace.html
- https://www.evm.codes
- https://github.com/ledgerwatch/erigon/blob/devel/core/vm/instructions.go
- https://github.com/ethereumbook/ethereumbook/blob/develop/13evm.asciidoc
- https://medium.com/coinmonks/ethereum-virtual-machine-evm-how-does-it-work-part-2-4198401d2a11
- https://hackernoon.com/getting-deep-into-evm-how-ethereum-works-backstage-ac7efa1f0015
- https://github.com/ledgerwatch/erigon/blob/devel/cmd/rpcdaemon/commands/trace_adhoc.go#L460

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

**Comment by banteg** [comment]
updated the gist with a compare mode, with the goal of this being a faster drop-in replacement for geth traces

Screen Shot 2022-07-24 at 07 20 01

**Comment by banteg** [comment]
i think i found a bug in erigon's vmTrace:
- `trace_replayTransaction, ['0x53546f6ac3d342842acfeeedeca5d66d7fda92acaa5bbc91ef94fd52b6f45a94', ['vmTrace']]`
- fails at this operation `VMOperation(pc=17605, cost=5, ex=VMExecutedOperation(used=298747, push=[], mem=None, store=None), sub=None, op='SMOD', idx='161-525-1253-663')`
- from my understanding `SMOD` consumes two items from the stack and puts back one, but here it doesn't put anything back

**Comment by banteg** [comment]
reverting the conversion of memory to hexbytes, just to show how much slower it is, consider these three methods:
1. construct hex bytes for each word: **1046.29 frames/s (baseline)** 
    `[HexBytes(memory.read_bytes(i, 32)) for i in range(0, len(memory), 32)]`
2. return a copy of memory bytes: **179451.29 frames/s (171.5x faster)**
    `memory.read_bytes(0, len(memory))`
3. avoid copying altogether with a memoryview: **244770.53 frames/s (233.9x faster)**
    `memory.read(0, len(memory))`

i think it's debatable whether we should return a memoryview, because it puts an assumption on a user, that they would process frames as they come in, otherwise the view might change.

i think we should go with the second option.

**Comment by banteg** [comment]
tests are failing because msgspec does not support python 3.7
https://jcristharif.com/msgspec/#installation

**Comment by banteg** [comment]
found a rare memory expansion inconsistency at 0xd884fade46358a2e093254b32db226b4c55435d8a83761e0bbd668a35531372a
Screen Shot 2022-07-24 at 11 28 57

**Comment by banteg** [comment]
added support for `trace_replayBlockTransactions`

Screen Shot 2022-07-25 at 09 51 27

**Comment by banteg** [comment]
moved demo here https://github.com/banteg/vmtrace

File: evm_trace/vmtrace.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    6: ["STATICCALL", "DELEGATECALL"],
+    7: ["CALL", "CALLCODE"]
+}
+# fmt: on
+POPCODES = {op: n for n, opcodes in POP_OPCODES.items() for op in opcodes}
+POPCODES.update({f"LOG{n}": n + 2 for n in range(0, 5)})
+POPCODES.update({f"SWAP{i}": i + 1 for i in range(1, 17)})
+POPCODES.update({f"DUP{i}": i for i in range(1, 17)})
+
+
```
Comment (fubuloubu) [inline evm_trace/vmtrace.py:32]: What value does this library have? Why not just use int in type hint

**Review by fubuloubu — APPROVED** [review]
Once this merges, please add an issue to remove the import error when we deprecate 3.7

File: evm_trace/vmtrace.py (hunk 1 above)
Comment (banteg) [reply evm_trace/vmtrace.py:32]: a fake uint256 type is needed to handle deserialization, since we need to convert hexstr to int

**Pull request merged**